samples/WinForms/Dx12Panel.cs

```diff
@@ -0,0 +1,868 @@
+// Copyright © Tanner Gooding and Contributors. Licensed under the MIT License (MIT). See License.md in the repository root for more information.
+
+// Ported from D3D12HelloWindow.h in https://github.com/Microsoft/DirectX-Graphics-Samples
+// Original source is Copyright © Microsoft. All rights reserved. Licensed under the MIT License (MIT).
+
+using System;
+using System.IO;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Windows.Forms;
+using TerraFX.Interop;
+using static TerraFX.Interop.D3D_FEATURE_LEVEL;
+using static TerraFX.Interop.D3D_PRIMITIVE_TOPOLOGY;
+using static TerraFX.Interop.D3D_ROOT_SIGNATURE_VERSION;
+using static TerraFX.Interop.D3D12;
+using static TerraFX.Interop.D3D12_COMMAND_LIST_TYPE;
+using static TerraFX.Interop.D3D12_DESCRIPTOR_HEAP_TYPE;
+using static TerraFX.Interop.D3D12_FENCE_FLAGS;
+using static TerraFX.Interop.D3D12_HEAP_FLAGS;
+using static TerraFX.Interop.D3D12_HEAP_TYPE;
+using static TerraFX.Interop.D3D12_INPUT_CLASSIFICATION;
+using static TerraFX.Interop.D3D12_PRIMITIVE_TOPOLOGY_TYPE;
+using static TerraFX.Interop.D3D12_RESOURCE_STATES;
+using static TerraFX.Interop.D3D12_ROOT_SIGNATURE_FLAGS;
+using static TerraFX.Interop.D3DCompiler;
+using static TerraFX.Interop.DXGI;
+using static TerraFX.Interop.DXGI_ADAPTER_FLAG;
+using static TerraFX.Interop.DXGI_FORMAT;
+using static TerraFX.Interop.DXGI_SWAP_EFFECT;
+using static TerraFX.Interop.Kernel32;
+using static TerraFX.Interop.Windows;
+using static TerraFX.Samples.DirectX.DXSampleHelper;
+
+
+namespace TerraFX.Samples.WinForms
+{
+    public unsafe class DX12Panel : Panel
+    {
+        #region local structs
+
+        private struct Vertex
+        {
+            public Vector3 Position;
+
+            public Vector4 Color;
+        }
+
+        #endregion local structs
+
+        #region data members
+
+        private float _aspectRatio;
+
+        // Adapter info
+        private readonly bool _useWarpDevice;
+
+        // Root assets path
+        private readonly string _assetsPath;
+
+        private const uint FrameCount = 2;
+
+        // Pipeline objects
+        private D3D12_VIEWPORT _viewport;
+        private RECT _scissorRect;
+        private IDXGISwapChain3* _swapChain;
+        private ID3D12Device* _device;
+        private readonly ID3D12Resource*[] _renderTargets;
+        private ID3D12CommandAllocator* _commandAllocator;
+        private ID3D12CommandQueue* _commandQueue;
+        private ID3D12RootSignature* _rootSignature;
+        private ID3D12DescriptorHeap* _rtvHeap;
+        private ID3D12PipelineState* _pipelineState;
+        private ID3D12GraphicsCommandList* _commandList;
+        private uint _rtvDescriptorSize;
+        private bool _isOnColorToggle;
```

Comment:
This isn't relevant to the `Panel` and is instead a feature of the sample, it shouldn't be here.

Reply:
hmm, so all the scene description and rendering should go to MainFormWinFormsDX12? Or better yet to a SceneModel class that then can also be used by a WPF sample and others?  Sounds good. Can that be done on a different PR (see next comment also)?

Reply:
I'm fine with it being handled in a follow up PR. Please log an issue to make sure it doesn't get forgotten however.

Reply:
created issue https://github.com/terrafx/terrafx.interop.windows/issues/61